[`pycodestyle`] Add an autofix for `E402`

PeterJCLaw opened the pull request:
## Summary

This implements a fixer for E402, as suggested at https://github.com/astral-sh/ruff/issues/6514.

This enables users to move imports up to the top of the file when they are found lower down. This is similar to isort's `float-to-top` logic, though not quite the same. As suggested by @ntBre on that issue, the expected path here is for people to enable unsafe-fixes to access this feature, potentially marking this fix "safe" in their config, rather than to implement an `float-to-top` option in the isort config.

Internally this means that there are separate fix edits for each import found to be out of place rather than a single one (the latter being how isort tends to work).

The current state of this is a _first draft_. It's working, I've done some manual testing and I have copied over a lot of the handling from the isort-like `organize_imports` function, though I may have missed things.

I'm also fairly new to rust and to this project so e.g: I don't know if e.g: the code is the best way to spell some of the concepts. I'm very happy to refactor things if there's better ways, including if e.g: we want to share more code between this and the isort crate -- I've deliberately not done that so far (purely to get to something which works).

This remains a bit work-in-progress, so I've included a test project I'm using for testing against though obviously that would be removed before this gets anywhere near merging. (I might also tidy the history a bit too)

Fixes https://github.com/astral-sh/ruff/issues/6514

## Test Plan

Manual testing so far, see various demo files currently committed. I'd like to add some tests and will look into that next. It looks like it's mostly snapshot based testing. Do we just add a new test function in `crates/ruff_linter/src/rules/pycodestyle/mod.rs`?

Noting them down so they're written, test cases:
- interaction with editor organise-imports logic -- with and without the E402 fix being accepted as safe
- variations on import spelling
- variations of other statements near the import, including on the same line (e.g: peruse the isort tests and pick relevant looking ones)

PeterJCLaw committed:
Update block output

PeterJCLaw committed:
Add reference project

PeterJCLaw committed:
More demo

PeterJCLaw committed:
WIP E402 fix

Essentially working, however doesn't preserve comments.
Untested in Jupyter notebook cells.

PeterJCLaw committed:
More demo details

PeterJCLaw committed:
First attempt to preserve comments

PeterJCLaw committed:
Enable more fixes in the test project

PeterJCLaw committed:
Some more imports to make it visible when editor import organisation runs

PeterJCLaw committed:
Include E402 fixing in editor import organising

Manual testing indicates that these still only get auto-fixed
if `extend-safe-fixes` contains `"E402"`.

PeterJCLaw committed:
Remove unused import

PeterJCLaw committed:
Remove comment-out code

PeterJCLaw committed:
Improve handling of lines around moved imports

This includes:
- not moving an import if there are other statements on the same
  line (copied over from the isort handling)
- picking up trailing comments (again copied from isort)
- removing but not double-inserting the trailing newline characters,
  so that fixes don't lead to lots of empty lines in the file

PeterJCLaw committed:
Remove debug files from previous approach

PeterJCLaw committed:
First pass at supporting fixing E402 in notebooks too

This doesn't seem to fully work -- the imports end up in the
cell above the one they came from (and sometimes further up).

PeterJCLaw committed:
Add fix title for E402 fixer

PeterJCLaw committed:
Update snapshot tests for E402 fixer

PeterJCLaw committed:
Comment about unsafe fixes in the organise imports action

In lieu of tests covering this it seems pertinent to record the
expected behaviour of including this fixer here.

PeterJCLaw committed:
Drop demo project now it's no longer needed

PeterJCLaw committed:
Add tests for e402 fixer

PeterJCLaw committed:
Ensure fixes for E402 are only moved to after future imports

__future__ imports must be the first statement in the file, so we
must ensure that and code moved to the "top" of the file remain
below any future imports already present.

PeterJCLaw commented:
Ok, have pulled out some shared parts of `add_import` and we're now using that, which looks like it avoids breaking `__future__` imports.

I still think there's a bug in `Insertion::start_of_file`, but I'm happy to pick that up separately as it's not trivial to either classify the issue nor fix it.

> It's not strictly required by our versioning guidelines, but I think I'd probably also add this fix in preview just to be safe since it feels a little risky to me.

I'll have a look at this when I get time, though I assume it won't significantly impact the rest of the PR which I believe is now ready for an initial review.

PeterJCLaw committed:
Merge branch 'main' into e402-fixer

PeterJCLaw committed:
Only enable E402 fixer in preview mode

This *moves* the test snapshots over to the new preview names so
the diff in this commit is easier to follow (and thus hopefully
the diff relative to the baseline is easier to see). A future
commit reinstates the non-preview snapshots.

PeterJCLaw committed:
Reinstate non-preview files

PeterJCLaw commented:
I've moved the fixer into being preview-only, I believe, and updated the tests.

Thanks for your help getting this far @ntBre; assuming that CI passes (tests pass locally) it would be great to get a review here when you have time.

PeterJCLaw commented:
@ntBre just wondering when you might have a chance to review this? I believe it's ready -- the fixer is preview only and there are tests for both the preview and non-preview cases.

ntBre commented on crates/ruff_server/src/server/api/requests/code_action_resolve.rs:
```diff
@@ -130,6 +130,13 @@ pub(super) fn organize_imports_edit(
     linter_settings.rules = [
         Rule::UnsortedImports,       // I001
         Rule::MissingRequiredImport, // I002
+        // Note: ModuleImportNotAtTopOfFile's fixes are unsafe. We include them
+        // here in order to match isort's behaviour and what we believe
+        // developers want. Since the fixes are unsafe, we're relying on this
+        // edit action not performing them unless the user has opted-in to these
+        // fixes in their settings (i.e: `extend-safe-fixes` in the
+        // `pyproject.toml` or similar).
+        Rule::ModuleImportNotAtTopOfFile, // E402
```
I was a bit worried about this initially, but if your comment is correct, it seems okay. Just double-checking, did you verify that this only works in an editor if `extend-safe-fixes` is set? It looks right to me from a quick skim, but it seems easy to verify empirically.

PeterJCLaw replied:
I believe I did early on, but I'll re-check and post back. Ideally I'd love to have a test for this but I couldn't easily find any tests which check this sort of thing. Do you have any suggestions for what such a test would look like (are there any examples I can borrow from or missed?)

PeterJCLaw replied:
Verified using VScode with ruff built from 671fbf2ae841d0b788df624da67377eca3a0d6b6 that the editor fix only applies when both `preview = true` and `extend-safe-fixes = ["E402"]` are set.
If there's a way add a test for this I'd be glad to do so.

ntBre commented on crates/ruff_linter/src/preview.rs:
```diff
@@ -327,6 +327,11 @@ pub(crate) const fn is_incorrect_dict_iterator_comprehension_enabled(
 }
 
+// https://github.com/astral-sh/ruff/pull/22212
+pub(crate) const fn is_import_float_to_top_enabled(settings: &LinterSettings) -> bool {
```
nit: I think I'd name this something more descriptive of the actual preview change, such as:


```suggestion
pub(crate) const fn is_e402_fix_enabled(settings: &LinterSettings) -> bool {
```

PeterJCLaw replied:
8582e57082c21f4ec6626ed0df49249bb4e0b234

ntBre commented on crates/ruff_linter/src/importer/mod.rs:
```diff
@@ -77,16 +77,21 @@ impl<'a> Importer<'a> {
+    }
+
+    /// Add a statement to the start of the file.
+    pub(crate) fn add_at_start(&self, text: &str, within_range: Option<TextRange>) -> Edit {
```
This API strikes me as a bit strange. Between this and the TODO comment, would it be helpful to skip the fix in notebooks for now? That would be totally fine with me if it would simplify the implementation.

PeterJCLaw replied:
66a1fe8729531e7f6c33fb01d54a60f524a24d2e improves the docstring. I've left the parameter since that's just a pass-through and I don't think should change anything in the callees.

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/rules/module_import_not_at_top_of_file.rs:
```diff
@@ -54,16 +60,63 @@ impl Violation for ModuleImportNotAtTopOfFile {
+
+        diagnostic.set_fix(Fix::unsafe_edits(
+            Edit::range_deletion(removal_range),
+            vec![edit],
```
nit: I think we can use an array or an `Option` to avoid allocating a vec:


```suggestion
            [edit],
```


```suggestion
            Some(edit),
```

PeterJCLaw replied:
4d73f92ef38149c6aab40b3421c37b30f479ef18

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__E402_E40.py.snap:
```diff
@@ -1,5 +1,6 @@
 ---
 source: crates/ruff_linter/src/rules/pycodestyle/mod.rs
+assertion_line: 82
```
I think this means you're on an old version of cargo-insta. Could you try updating and see if these go away?

PeterJCLaw replied:
1a4b05313472cec207e6233c9fc8bad442e367ec

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/rules/module_import_not_at_top_of_file.rs:
```diff
@@ -1,9 +1,13 @@
+use crate::{Edit, Fix, FixAvailability, Violation};
 
-use crate::Violation;
 use crate::checkers::ast::Checker;
```
nit: Looks like there's an extra blank here between some of the `crate::` imports.

PeterJCLaw replied:
0a34521dcd7b4cb2ccbb50d7566118c588350019

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/mod.rs:
```diff
@@ -46,6 +46,14 @@ mod tests {
+    #[test_case(
+        Rule::ModuleImportNotAtTopOfFile,
+        Path::new("E402_shebang_docstring_and_future.py")
+    )]
```
I think it's actually okay not to test these with preview disabled since the preview behavior is the purely additive fix change. The main thing we want to assert here is that the fix is _not_ offered, which will be covered by the existing E402 tests here.

PeterJCLaw replied:
Sure. I assume normal practise when moving something out of preview is to copy over the tests?
~08bf1899d25f781ab1aee951c7f910615ac120b2~ 53e28182f583e8a9f4c1ae14d65314726b797d3f

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__preview__E402_E402_comments.py.snap:
```diff
@@ -0,0 +1,176 @@
+25 | | )
+   | |_^
+   |
+help: Move module level imports to top of file
```
One potential issue here is that moving multiple imports reverses the order, if I'm following correctly. We ran into something similar recently with `I002` (https://github.com/astral-sh/ruff/issues/20811), but the fix here wouldn't be so easy since we don't know all of the imports we're moving up front. I'm okay with this limitation because I don't see a good way around it, at least with the current structure of the rule, but we should probably document it somewhere.

PeterJCLaw replied:
Ooh, good catch. I expect that a lot of people running this will also be using the isort rules, but given this is no longer coupled to those we can't rely on it.

I'm not immediately finding notes about linters, but in lots of cases there are notes about lints and their effects on the rule pages, so that seems a good place. I'll add something to https://docs.astral.sh/ruff/rules/module-import-not-at-top-of-file/. (I see you also suggesting adding something there about fix safety)

PeterJCLaw replied:
Hopefully 6271fc947e32ef85f18b3701e25841050e8b5296 covers this?

ntBre commented on crates/ruff_linter/resources/test/fixtures/pycodestyle/E402_docstring.py:
```diff
@@ -0,0 +1,8 @@
+import os
+
+# Prevent F403
+os.path
```
Not really a big deal, but what does this comment mean? I don't think [undefined-local-with-import-star (F403)](https://docs.astral.sh/ruff/rules/undefined-local-with-import-star/#undefined-local-with-import-star-f403) should be triggered here, and `E402` is probably the only rule enabled, at least with our typical test setup.

PeterJCLaw replied:
Hrm, this might have been derived from a file I used for manual testing in an IDE. I'll check the test works without it; happy to remove if so.

PeterJCLaw replied:
671fbf2ae841d0b788df624da67377eca3a0d6b6

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__preview__E402_E40.py.snap:
```diff
@@ -0,0 +1,126 @@
+--------------------------------------------------------------------------------
+56 |
+57 | import foo
+58 | #: E402
+   - import foo
```
It's slightly unfortunate to leave the comment hanging, but I don't think we can easily reuse the logic from the isort rules that tries to move comments too.

PeterJCLaw replied:
Yeah, I've tried to copy over what I could. As far as I can tell the isort rules don't actually support it either (though `isort` itself does). I hope it's ok to leave this and let both be addressed together.

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/snapshots/ruff_linter__rules__pycodestyle__tests__preview__E402_E402.ipynb.snap:
```diff
@@ -0,0 +1,80 @@
+20 | __some__magic = 1
+21 |
+22 | import c
+23 + import no_ok
```
I think you know about this since there's already a TODO comment, but it looks like this case is the one place where the import moves across cell boundaries. Again, I'm leaning toward excluding notebooks for now, just noting this for future reference.

PeterJCLaw replied:
Looking at notebooks again, I'd actually hoped to get them working here. Though agree that it'd be simpler to drop them for now.

In an editor (VSCode at least) running the fix action in a notebook cell does work correctly currently, which suggests to me that there's something happening in the interaction with a file full of cells instead of whatever the editor interaction is. My first thought was that this was a cascading error since we're processing top to bottom, but that doesn't seem to fully explain what's happening here.

Anyway, 33b5efb9d694f56d279325bd199e9527fa0c934a disables the fixer for notebooks. (Though it looks like it still works in editor sessions :shrug:)

ntBre commented on crates/ruff_linter/src/rules/pycodestyle/rules/module_import_not_at_top_of_file.rs:
We should add a `## Fix safety` section to the end of the docs noting that the fix is always unsafe because reordering imports can change runtime behavior. That's hopefully kind of obvious to readers but good to have the section anyway.

PeterJCLaw replied:
6271fc947e32ef85f18b3701e25841050e8b5296

KMohZaid replied:
> [6271fc9](https://github.com/astral-sh/ruff/commit/6271fc947e32ef85f18b3701e25841050e8b5296)

I think it can be clarified as for example some library ask for specific code to be executed before they are imported for example 
```
import gi
gi.require_version("Pango", "1.0")
gi.require_version("PangoCairo", "1.0")
from gi.repository import Pango, PangoCairo  
```

This is indented  else it will throw error if we don't specify required version before importing 

So this can be like clarified in docs as some imports after lines of code can be intentional which formatter can't tell so this is unsafe fix. And to avoid formatting in such case add comment like `# noqa: E402` at end of line 

PS : idk if this will make formatter to ignore sorting this line but I think it will make ruff linter don't show diagnostic for given line and current format code that you implement is only moving line that linter specify as E402 right?

ntBre reviewed (commented):
Thank you for all of your work here and sorry again for the delay!

I only had a few minor comments and a suggestion to drop notebook support for now to simplify some things. I noticed a couple of other issues that I think are probably okay for now, but I'm curious to get your thoughts.

PeterJCLaw committed:
Merge branch 'main' into e402-fixer

PeterJCLaw committed:
Rebuild snapshots for newer cargo-insta

PeterJCLaw committed:
Group imports

PeterJCLaw committed:
Avoid allocation we don't need

PeterJCLaw committed:
Clarify helper function name

PeterJCLaw committed:
Add notes on fix safety

PeterJCLaw committed:
Remove redundant code from snapshot tests

PeterJCLaw committed:
Improve comment

PeterJCLaw committed:
Drop attempted support for notebooks

It doesn't work currently, so it's simpler to drop it for now.

PeterJCLaw committed:
No need to run these new fixer tests non-preview

While the fixer is in preview, adding these in the non-preview
list adds little value. They should however be moved over once
the fixer matures from preview.

PeterJCLaw commented:
@ntBre gentle ping on review here, I believe I've responded to/addressed all the comments.

PeterJCLaw committed:
Merge branch 'main' into e402-fixer

PeterJCLaw committed:
Update snapshots following #26161

ntBre commented:
Thank you! And sorry for the delay. This looks good to me now. My very last suggestion is around the importer API again. What do you think about something like this?

<details><summary>Patch</summary>

```diff
diff --git a/crates/ruff_linter/src/importer/mod.rs b/crates/ruff_linter/src/importer/mod.rs
index 95958c6640..3387a79a96 100644
--- a/crates/ruff_linter/src/importer/mod.rs
+++ b/crates/ruff_linter/src/importer/mod.rs
@@ -17,7 +17,8 @@ use ruff_python_parser::Parsed;
 use ruff_python_semantic::{
     ImportedName, MemberNameImport, ModuleNameImport, NameImport, SemanticModel,
 };
-use ruff_python_trivia::textwrap::indent;
+use ruff_python_trivia::{PythonWhitespace, textwrap::indent};
+use ruff_source_file::LineRanges;
 use ruff_text_size::{Ranged, TextRange, TextSize};
 
 use crate::cst::matchers::{match_aliases, match_import_from, match_statement};
@@ -77,22 +78,22 @@ impl<'a> Importer<'a> {
             // Insert after the last top-level import.
             Insertion::end_of_statement(stmt, self.source, self.stylist).into_edit(&required_import)
         } else {
-            self.add_at_start(&required_import, None)
+            self.add_at_start(&required_import)
         }
     }
 
-    /// Add a statement to the start of the file, potentially limited to a text range.
-    ///
-    /// The latter is useful for operating within notebooks.
-    pub(crate) fn add_at_start(&self, text: &str, within_range: Option<TextRange>) -> Edit {
-        // Check if there are any future imports that we need to respect
-        if let Some(last_future_import) = self.find_last_future_import(within_range) {
-            // Insert after the last future import
+    /// Add an existing import statement to the start of the file.
+    pub(crate) fn add_import_at_start(&self, import: &Stmt) -> Edit {
+        let range = TextRange::new(import.start(), self.source.line_end(import.end()));
+        self.add_at_start(self.source[range].trim_whitespace())
+    }
+
+    fn add_at_start(&self, text: &str) -> Edit {
+        if let Some(last_future_import) = self.find_last_future_import() {
             Insertion::end_of_statement(last_future_import, self.source, self.stylist)
                 .into_edit(text)
         } else {
-            // Insert at the start of the file.
-            Insertion::start_of_file(self.python_ast, self.source, self.stylist, within_range)
+            Insertion::start_of_file(self.python_ast, self.source, self.stylist, None)
                 .into_edit(text)
         }
     }
@@ -542,22 +543,8 @@ impl<'a> Importer<'a> {
     }
 
     /// Find the last `from __future__` import statement in the AST.
-    fn find_last_future_import(&self, within_range: Option<TextRange>) -> Option<&'a Stmt> {
-        // When operating within notebook cells we might only want to look at a
-        // range of the AST, yet can still encounter future imports at the top
-        // of a cell.
-        let full_body = self.python_ast;
-        let mut body = within_range
-            .map(|range| {
-                let start = full_body.partition_point(|stmt| stmt.start() < range.start());
-                let end = full_body.partition_point(|stmt| stmt.end() <= range.end());
-
-                &full_body[start..end]
-            })
-            .unwrap_or(full_body)
-            .iter()
-            .peekable();
-
+    fn find_last_future_import(&self) -> Option<&'a Stmt> {
+        let mut body = self.python_ast.iter().peekable();
         let _docstring = body.next_if(|stmt| ast::helpers::is_docstring_stmt(stmt));
 
         body.take_while(|stmt| {
diff --git a/crates/ruff_linter/src/rules/pycodestyle/rules/module_import_not_at_top_of_file.rs b/crates/ruff_linter/src/rules/pycodestyle/rules/module_import_not_at_top_of_file.rs
index 81886da596..0112808f32 100644
--- a/crates/ruff_linter/src/rules/pycodestyle/rules/module_import_not_at_top_of_file.rs
+++ b/crates/ruff_linter/src/rules/pycodestyle/rules/module_import_not_at_top_of_file.rs
@@ -1,6 +1,5 @@
 use ruff_macros::{ViolationMetadata, derive_message_formats};
 use ruff_python_ast::{PySourceType, Stmt};
-use ruff_python_trivia::PythonWhitespace;
 use ruff_source_file::LineRanges;
 use ruff_text_size::{Ranged, TextRange};
 
@@ -105,18 +104,10 @@ pub(crate) fn module_import_not_at_top_of_file(checker: &Checker, stmt: &Stmt) {
             return;
         }
 
-        let range = stmt.range();
+        let edit = checker.importer().add_import_at_start(stmt);
 
-        // Include comments but not the trailing newline (so we don't insert an extra newline).
-        let text_range = TextRange::new(range.start(), locator.line_end(range.end()));
-
-        let edit = checker
-            .importer()
-            .add_at_start(checker.source()[text_range].trim_whitespace(), None);
-
-        // Include comments *and* the trailing newline, so that we do remove the whole line.
-        let removal_range =
-            TextRange::new(text_range.start(), locator.full_line_end(text_range.end()));
+        // Include trailing comments and the newline in the removal.
+        let removal_range = TextRange::new(stmt.start(), locator.full_line_end(stmt.end()));
 
         diagnostic.set_fix(Fix::unsafe_edits(
             Edit::range_deletion(removal_range),
```

</details>

This reverts the now-unused `within_range` changes and makes the `str` version of `add_at_start` private, only exposing the `add_import_at_start` helper, which also simplifies the calling side a bit too.

PeterJCLaw committed:
Merge branch 'main' into e402-fixer

ntBre committed:
Simplify import insertion logic dropping partial notebook support

The notebook support added earlier on this branch doesn't fully
work, though did add a little complexity. This commit removes
most of it, leaving the code simpler without losing functionality
which is known working.

ntBre reviewed (approved):
Thank you!

PeterJCLaw commented:
Thanks for your help with this @ntBre :)
